### ISSUE TITLE
False-Positive when property getter and setter heve different types

### ISSUE THREAD

**Issue by saroad2** [python/mypy#9351]
When a property setter and getter have different signatures, false-positive error arises. Consider the following example:

```python
from typing import Optional


class A:
    __foo: Optional[int] = None

    @property
    def foo(self) -> int:
        if self.__foo is None:
            raise ValueError("foo is None!")
        return self.__foo

    @foo.setter
    def foo(self, foo: Optional[int]) -> None:
        self.__foo = foo

    def reset(self) -> None:
        self.foo = None  # <- mypy error here
```

As you see, class `A` has a property named `foo` which returns an integer. However, one can unset `foo` to be `None` if it has no value. If a user tries to call `foo` when it is unset, a `ValueError` exception arises.

When running *mypy* in **strict** mode on that code, I get the following error:
```
src\mypy_bug.py:18: error: Incompatible types in assignment (expression has type "None", variable has type "int")
Found 1 error in 1 file (checked 1 source file)
```

But as we see, the signature of the property setter states that `None` **can** be received as an input.

It seems like a bug in *mypy*.

**Comment by hauntsaninja** [comment]
Duplicate of https://github.com/python/mypy/issues/3004
